File: Cpp/string/z_algorithm.cpp
#include <string>
#include <vector>

std::vector<int> z_algorithm(const std::string& s) {
	int len = (int)s.size();
	std::vector<int> res(s.size());
	res[0] = s.size();
	int i = 1, j = 0;
	while (i < s.size()) {
		while (i + j < s.size() && s[i + j] == s[j]) j++;
		res[i] = j;
		if (j == 0) {
			i++;
			continue;
		}
		int k = 1;
		while (i + k < s.size() && j > k + res[k]) {
			res[i + k] = res[k];
			k++;
		}
		i += k;
		j -= k;
	}
	return res;
}

File: Python/string/z_algorithm.py
# Z-Algorithm
def z_algorithm(s: str) -> list:
	str_len = len(s)
	res = [0] * str_len
	res[0] = str_len
	i, j = 1, 0
	while i < str_len:
		while i + j < str_len and s[i + j] == s[j]:
			j += 1
		res[i] = j
		if j == 0:
			i += 1
			continue
		k = 1
		while i + k < str_len and j > res[k] + k:
			res[i + k] = res[k]
			k += 1
		i += k
		j -= k
	return res